File: desktop/src/services/api/transports/http/audio.ts
import type { HttpClient } from '../../http-client';

export interface ElevenLabsVoice {
  voice_id: string;
  name: string;
  label: string;
}

export interface TranscribeResponse {
  ok: boolean;
  transcript: string;
  provider?: string | null;
}

export interface SpeakResponse {
  ok: boolean;
  data_url: string;
  mime_type: string;
  provider?: string | null;
}

export interface ElevenLabsVoicesResponse {
  available: boolean;
  voices: ElevenLabsVoice[];
}

export interface AudioTransport {
  transcribe(dataUrl: string, mimeType?: string): Promise<TranscribeResponse>;
  speak(text: string): Promise<SpeakResponse>;
  getElevenLabsVoices(): Promise<ElevenLabsVoicesResponse>;
}

export function makeAudioTransport(client: HttpClient): AudioTransport {
  return {
    transcribe: (dataUrl, mimeType) =>
      client.post<TranscribeResponse>('/desktop/api/audio/transcribe', {
        data_url: dataUrl,
        ...(mimeType ? { mime_type: mimeType } : {}),
      }),

    speak: (text) =>
      client.post<SpeakResponse>('/desktop/api/audio/speak', { text }),

    getElevenLabsVoices: () =>
      client.get<ElevenLabsVoicesResponse>('/desktop/api/audio/elevenlabs/voices'),
  };
}

File: desktop/sidecar/daemon/schemas/audio.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class AudioTranscriptionRequest(BaseModel):
    data_url: str
    mime_type: Optional[str] = None


class AudioTranscriptionResponse(BaseModel):
    ok: bool
    transcript: str
    provider: Optional[str] = None


class TTSSpeakRequest(BaseModel):
    text: str


class TTSSpeakResponse(BaseModel):
    ok: bool
    data_url: str
    mime_type: str
    provider: Optional[str] = None


class ElevenLabsVoice(BaseModel):
    voice_id: str
    name: str
    label: str


class ElevenLabsVoicesResponse(BaseModel):
    available: bool
    voices: list[ElevenLabsVoice]
